### ISSUE TITLE
[Tracking] azk.io kubernetes deployment

### ISSUE THREAD

**Issue by paralin** [r3stack/r3stack#5]
I mentioned in the deployment thread on Meatier the possibility of developing an easy way to deploy to Kubernetes via Azk.

I'm now working towards implementing a deployment image for Kubernetes via azk.io.

I will post some regular updates here. Please just close it if you think it's not relevant to r3stack to have this here.

**Comment by paralin** [comment]
Will be implemented here: https://github.com/paralin/docker-deploy-kubernetes

**Comment by paralin** [comment]
Actually this is going to be realllly promising for a deployment tool to Kubernetes. Built in rollbacks, "azk shell" can just alias to "kubectl exec -it *container* bash"... Nice.

**Comment by Bartekus** [comment]
I find anything that expands knowledge and helps people grow and achieve their dreams/objectives are most appropriate so in fact I encourage anything and everything of this nature @paralin . Thank you for your initiative and contributions!

**Comment by paralin** [comment]
Okay, after a long night of hacking.. It's all but done. Just need to implement image push (there will be some question of where the images should go, though).

Completed:
- System analyzes Azkfile.js, converts it to basic JSON.
- JSON from azk2json converted into Kubernetes yaml configurations using reasonable best practices
- Existing cluster is analyzed for differences from the suggested model. Steps to actuate to target model are generated and printed in a human readable form.
- Actions are performed to bring cluster up to target state.

Sample output:

First run: http://sprunge.us/PbUa

Second run: http://sprunge.us/ODhf

**Comment by Bartekus** [comment]
Outstanding work @paralin, being that deployment is always big part of the development equation, your work goes a long way in giving options and expanding the choices that programmers will be able to have and make respectively!

**Comment by paralin** [comment]
The azk developers seem to want me to use their ansible/ssh based deploy base image for the kubernetes base. this would never work. hopefully I can convince them to allow any image to be dropped in instead.

If not I think the entire azk development platform mentality (keeping everything centralized and dependent on Azukiapp) is flawed and it might be better to abandon that approach completely.

It also doesn't make too much sense to deploy via ssh and traditional ansible methods when you're working within Docker containers anyway. Overall it just seems like a really poorly designed system.

**Comment by Bartekus** [comment]
@paralin I agree, a solution which is flexible and optimal is the best solution and lack of willingness adapt or adjust (while certainly is within their rights to do so), is not something we have to or should abide by. Nothing forced is ever good in the long run. Beside being that azk is docker based, indeed it makes little sense to not use the available docker potential to its full capability. With meatier we are also adapting similar half-way path and being that it's sister-project, we'll be transitioning the found solutions between the two, trying to accomplish the most optimal and balanced path.
